lib/vector/Vlib/copy.c

```diff
@@ -567,6 +574,10 @@ int Vect__copy_areas(struct Map_info *In, int field, struct Map_info *Out)
             if (0 > V2__write_area_sfa(Out, (const struct line_pnts **)Points,
                                        nisles + 1, Cats)) {
                 G_warning(_("Writing area %d failed"), area);
+                Vect_destroy_cats_struct(Cats);
+                for (i = 0; i < nparts_alloc; i++)
+                    Vect_destroy_line_struct(Points[i]);
+                G_free(Points);
```

Comment:
Let us declare a `int ret;`, initialised to `0` at the top of the function.

Here set `ret = -1`, and `goto free_exit;`

The above also needs to be done at around L577, in `#ifdef HAVE_POSTGRES`.

Create a goto label `free_exit:` at line 588 before the comment `/* free allocated space for isles */`.